modules: correctly detect explicit installation

xdelaruelle opened the pull request:
When generating modulefile, correctly detect software installation asked by user as explicit installation.

Explicit installation status were previously fetched from database record of spec, which was only set after modulefile generation.

Code is updated to pass down the explicit status of software installation to the object that generates modulefiles.

Fixes #34730.
Fixes #12105.

xdelaruelle committed:
modules: correctly detect explicit installation

When generating modulefile, correctly detect software installation asked
by user as explicit installation.

Explicit installation status were previously fetched from database
record of spec, which was only set after modulefile generation.

Code is updated to pass down the explicit status of software
installation to the object that generates modulefiles.

Fixes #34730.
Fixes #12105.

xdelaruelle commented on lib/spack/spack/installer.py:
```diff
@@ -1260,6 +1260,10 @@ def _install_task(self, task):
 
+        # Injecting information to know if this installation request is the root one
+        # to determine in BuildProcessInstaller whether installation is explicit or not
+        install_args["is_root"] = task.is_root
```
I am not sure of this part, but I found that adding `explicit` argument down the `spack.build_environment.start_build_process` way would imply too much changes.

haampie replied:
There should be a prop `explicit` already, which should be respected

xdelaruelle replied:
@haampie `explicit` prop is about the overall installation, not the specific one treated by the `start_build_process`. So when installing an implicit dependency, `install_args["explicit"]` will be `True` as the installation of the dependent module has been asked by user.

We need to mix this `explicit` information to the `is_root` one (is this the software asked by user or one of its dependency) to get the accurate implicit state.

haampie replied:
Right. I meant the task's prop, which is defined as `self.is_root and self.request.install_args.get("explicit", True)` (but not available in `def build_process`).

I would personally not duplicate this logic in different places but consolidate ahead of time. So, can't you update `args_for_build_process["explicit"] = task.explicit`? (Edit: I guess not immediately, since all tasks seem to hold a reference to the same dict :|)

haampie replied:
I'm sorry you have to deal with this :laughing: looking at the code it's generally a PITA. It was a mistake to pass effectively command line arguments all the way to an individual package installs, instead of transforming them s.t. things like `explicit` were package specific.

xdelaruelle replied:
Do you mean in `def _install_task` to update `install_args["explicit"]` to make it equal to `task.explicit`?

I added another item to the dict (`is_root`) in my initial code, as I thought there may be some underlying part requiring this `explicit` item to be at its initial value.

But if this is the way you suggest, no problem I will update the PR to follow your suggested approach.

haampie replied:
Thinking about it more the PR is probably good as is. I didn't realize every task holds a reference to the same `install_args` hashmap, so mutating it is a bad idea, and changing it s.t. it's task-specific is maybe beyond this PR.

xdelaruelle commented:
I have added a basic test, to check that `explicit` argument passed to modulefile writer object gives the correct `excluded` state when `exclude_implicit` option is enabled.

haampie commented on lib/spack/spack/installer.py:
```diff
@@ -1879,6 +1883,9 @@ def __init__(self, pkg, install_args):
         self.verbose = install_args.get("verbose", False)
 
+        # whether installation was explicitly requested by the user
+        self.explicit = install_args.get("is_root", False) and install_args.get("explicit", False)
```
```suggestion
        self.explicit = install_args.get("is_root", False) and install_args.get("explicit", True)
```

should duplicate `task.explicit`

haampie commented on lib/spack/spack/modules/common.py:
```diff
@@ -428,12 +428,17 @@ class BaseConfiguration(object):
         self.name = module_set_name
+        # Software installation has been explicitly asked (get this
+        # information from spec if not provided as argument)
+        if explicit is None:
```
Under what conditions is this branch followed?

xdelaruelle replied:
When we query existing modulefiles with `get_module` def (defined in [lib/spack/spack/modules/common.py](https://github.com/spack/spack/blob/develop/lib/spack/spack/modules/common.py#L400-L403)), we need to query the database to learn from it if the module is implicit or explicit.

I will update the comment to clarify.

xdelaruelle committed:
Set explicit in BuildProcessInstaller with task.explicit logic

xdelaruelle committed:
Clarify when explicit is None in BaseConfiguration

A value for the explicit argument has to be set when creating a new
installation, but for operations on existing installation, this value is
retrieved from database. Such operations are: module rm, module refresh,
module setdefaults or when get_module function is used.

Update on the way tests that mimics an installation, thus explicit
argument has to be set under such situation.

xdelaruelle committed:
Remove unused args on exclude_implicits tests

haampie commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -2023,7 +2023,7 @@ def install_root_node(spec, allow_root, unsigned=False, force=False, sha256=None
         tty.msg('Installing "{0}" from a buildcache'.format(spec.format()))
         extract_tarball(spec, download_result, allow_root, unsigned, force)
-        spack.hooks.post_install(spec)
+        spack.hooks.post_install(spec, False)
```
This is probably OK. It's also called through `spack buildcache install`, but I think that's more of a legacy command, so let's ignore it.

haampie reviewed (approved):
Thanks!
